core/src/main/java/org/apache/calcite/rel/mutable/MutableRels.java

```diff
@@ -173,6 +173,23 @@ public RexNode get(int index) {
         .collect(Collectors.toList());
   }
 
+  /**
+   * Construct expression list of Project by the given fields of the input.
+   */
```

Comment:
Seems this method only converts the row type for `RexInputRef`, why this is taken into consideration specially ?
BTW I can pass the test you provided in MaterializationTest.java without this change.